### ISSUE TITLE
GitHub hourly rate limit

### ISSUE THREAD

**Issue by Beanow** [sourcecred/sourcecred#1352]
So I tried out a repo with a fair amount of data to load. passbolt/passbolt_api
As part of testing out a blacklist change, see https://github.com/sourcecred/sourcecred/issues/1256#issuecomment-526135193
After 40 minutes, the loading failed.

```
  GO   load-@passbolt
  GO   github/passbolt/passbolt_api
You've exceeded your hourly GitHub rate limit.
You'll need to wait until it resets.
{"type":"RATE_LIMIT_EXCEEDED","retry":false,"error":{"errors":[{"type":"RATE_LIMITED","message":"API rate limit exceeded"}]}}
```

In my case though I was running this from a container, as with the blacklist changes I expected to need to be sure the cache isn't causing issues. So I didn't persist it.

This failure mode caused those 40m to in the end be an unfortunate spamming of the GitHub API that I'll need to repeat to get any results.

We might want to handle this a bit more gracefully. For example baking an understanding of the rate limiting into the GitHub plugin and pause loading of data with some warning messages we're doing so.

**Comment by decentralion** [comment]
Would you like to see a behavior change where when it hits a API rate limit, it prints a message, pauses for an hour, and then resumes?

**Comment by Beanow** [comment]
> Would you like to see a behavior change where when it hits a API rate limit, it prints a message, pauses for an hour, and then resumes?

An hour - elapsed time, or just attempt a resume every 1-2 minutes.
But yes resume basically.

**Comment by wchargin** [comment]
No need to guess how long to sleep; [just ask:][rl]

```graphql
{
  rateLimit {
    remaining
    resetAt
  }
}
```

```json
{
  "data": {
    "rateLimit": {
      "remaining": 4999,
      "resetAt": "2019-09-02T16:09:59Z"
    }
  }
}
```

Querying the `rateLimit` struct itself does not count against your rate
limit, and it’s possible that you can even issue this query once your
rate limit has expired (though I haven’t tested this).

[rl]: https://developer.github.com/v4/query/#ratelimit

**Comment by wchargin** [comment]
> In my case though I was running this from a container […] So I didn't
> persist it. […] This failure mode caused those 40m to in the end be an
> unfortunate spamming of the GitHub API that I'll need to repeat to get
> any results.

As to the lost work—perhaps you could configure your container to use
merely a fresh cache, mounted at a new directory on the host filesystem,
rather than no cache? Rate limiting, network errors, nondeterministic
server-side flukes, and the like do happen fairly frequently; the
ability to resume partial downloads is a primary reason that the Mirror
module exists in the first place. At the end of the day, there’s not
much we can do in the way of workarounds when all checkpointing is
explicitly disabled.

**Comment by jakubgs** [comment]
Any plans for adding any handling for this? Some GitHub organizations are so big that one cannot get through a full `load` run without hitting the `RATE_LIMIT_EXCEEDED` error.

Either some backoff should be implemented, or even better, a configurable delay between calls, so as to be more respectful of the GitHub API. I don't mind if the job runs for 24 hours instead of 3 by adding big delays between calls, as long as I can get all the data in one run without errors.

**Comment by decentralion** [comment]
At the moment, I am not prioritizing this. The reason is that it isn't a blocker for the [CredSperiment](https://discourse.sourcecred.io/t/about-the-credsperiment/181) and [I'm prioritizing the CredSperiment before features for external users](https://discourse.sourcecred.io/t/make-the-github-plugin-robust/256/2?u=decentralion). 

I wrote a tracking initiative for the [myriad usability problems with the GitHub plugin](https://discourse.sourcecred.io/t/make-the-github-plugin-robust/256), and as soon as I have the ability to set cred bounties on initiatives, I will add a substantial bounty to that initiative. So if you decide to take a stab at fixing this, you will (retroactively) get a bunch of cred.

**Comment by Beanow** [comment]
Hey @jakubgs :smile:

Some quality-of-life improvements in the loading process would be wonderful. And yeah as @decentralion fairly points out. We don't strictly need it in the short term to run CredSperiment. But hitting a program crash when hitting the rate limit isn't a great experience.

Another one that comes to mind is having a progress indicator of sorts. When we're already talking a load that takes more than an hour to complete, some progress feedback would be much appreciated.

Is this something you would be able to help with @jakubgs?

**Comment by jakubgs** [comment]
I'm checking if I can do it as part of my duties as sysadmin for Status.im. We'll see.

**Comment by befitsandpiper** [comment]
This issue was resolved in release 0.7.1
